lib/base-compiler.js

```diff
@@ -895,7 +957,157 @@ export class BaseCompiler {
+            ldPath: this.getSharedLibraryPathsAsLdLibraryPaths(key.libraries),
+            args: key.executionParameters.args || [],
+            stdin: key.executionParameters.stdin || '',
+        };
+
+        const cacheKey = this.getCmakeCacheKey(key, files);
+
+        const dirPath = await this.newTempDir();
+        const outputFilename = this.getExecutableFilename(dirPath, this.outputFilebase);
+
+        let fullResult = await this.loadPackageWithExecutable(cacheKey, dirPath);
+        if (!fullResult) {
+            const filesToWrite = [];
+            filesToWrite.push(fs.writeFile(path.join(dirPath, 'CMakeLists.txt'), cacheKey.source));
+            for (let file of files) {
+                filesToWrite.push(fs.writeFile(this.getExtraFilepath(dirPath, file.filename), file.contents));
+            }
+
+            const execParams = this.getDefaultExecOptions();
+            execParams.customCwd = dirPath; //path.join(dirPath, 'build');
+
+            const makeExecParams = this.createCmakeExecParams(execParams, dirPath, libsAndOptions);
+
+            await Promise.all(filesToWrite);
+
+            fullResult = {
+                buildsteps: [],
+            };
+
+            await this.setupBuildEnvironment(cacheKey, dirPath);
+
+            let toolchainparam = '';
+            if (this.toolchainPath) {
+                toolchainparam = `-DCMAKE_CXX_COMPILER_EXTERNAL_TOOLCHAIN=${this.toolchainPath}`;
+            }
+
+            await this.doBuildstepAndAddToResult(fullResult, 'cmake', this.env.ceProps('cmake'),
+                [toolchainparam, '.'], makeExecParams);
+            await this.doBuildstepAndAddToResult(fullResult, 'make', 'make',
+                [], execParams);
+
+            fullResult.result = {
+                dirPath,
+                okToCache: true,
+            };
+
+            const [asmResult] = await this.checkOutputFileAndDoPostProcess(
+                fullResult.result, outputFilename, cacheKey.filters);
+            fullResult.result = asmResult;
+
+            await this.storePackageWithExecutable(cacheKey, dirPath, fullResult);
+        } else {
+            fullResult.fetchedFromCache = true;
+
+            delete fullResult.code;
+            delete fullResult.inputFilename;
+            delete fullResult.dirPath;
+            delete fullResult.executableFilename;
+        }
+
+        const optOutput = undefined;
+        await this.afterCompilation(fullResult.result, false, cacheKey, [], cacheKey.tools, cacheKey.backendOptions,
+            cacheKey.filters, libsAndOptions.options, optOutput);
+
+        if (this.compiler.supportsExecute && doExecute) {
+            fullResult.execResult = await this.runExecutable(outputFilename, executeParameters);
+            fullResult.didExecute = true;
+        }
+
+        return fullResult;
+    }
+
+    getExtraFilepath(dirPath, filename) {
+        // todo: sanitize filename first
+        return path.join(dirPath, filename);
```

Comment:
Anyone with a suggestion on how to do this?

Would https://www.npmjs.com/package/sanitize-filename be good enough for this?

Reply:
The main thing this should be able to do is not get out of the `dirPath` in any way.

Reply:
Hot take and one without reading the rest of the PR (yet) - If the actual filename is unimportant we can use the hash of the filename instead? Proably dumb....

Reply:
Unfortunately, the point of these files is that you can refer to them from either the CMakeLists.txt file or from the users' code. So that wouldn't work, I think.

Reply:
Got it. Ensuring it's only `[a-zA-Z0-9_.]+` is probably "good enough" and/or the library you mentioned too.

Reply:
calling path.normalize on the user controlled segment before joining might do it (although please check me on this).

Alternatively, call path.normalize after joining and ensure it starts with `dirPath + path.sep`

Reply:
the latter would be quite useful if we wanna support subfolders, for now I got the regex working